cli/tools/registry/pm.rs

```diff
@@ -25,22 +28,149 @@ use crate::file_fetcher::FileFetcher;
+              bail!("Malformed \"imports\" configuration: {err}")
+            }
+          }
+        } else {
+          Ok(Default::default())
+        }
+      }
+      ConfigFile::Npm(npm, ..) => {
+        Ok(npm.dependencies.clone().unwrap_or_default())
+      }
+    }
+  }
+
+  fn fmt_options(&self) -> FmtOptionsConfig {
+    match self {
+      ConfigFile::Deno(deno) => deno
+        .to_fmt_config()
+        .ok()
+        .flatten()
+        .map(|f| f.options)
+        .unwrap_or_default(),
+      ConfigFile::Npm(_, config) => config.clone().unwrap_or_default(),
+    }
+  }
+
+  fn imports_key(&self) -> &'static str {
+    match self {
+      ConfigFile::Deno(_) => "imports",
+      ConfigFile::Npm(..) => "dependencies",
+    }
+  }
+
+  fn file_name(&self) -> &'static str {
+    match self {
+      ConfigFile::Deno(_) => "deno.json",
+      ConfigFile::Npm(..) => "package.json",
+    }
+  }
+
+  fn is_npm(&self) -> bool {
+    matches!(self, Self::Npm(..))
+  }
+
+  /// Get the preferred config file to operate on
+  /// given the flags. If no config file is present,
+  /// creates a `deno.json` file - in this case
+  /// we also return a new `CliFactory` that knows about
+  /// the new config
+  async fn from_flags(flags: Flags) -> Result<(Self, CliFactory), AnyError> {
+    let factory = CliFactory::from_flags(flags.clone())?;
+    let options = factory.cli_options().clone();
+
+    match (options.maybe_config_file(), options.maybe_package_json()) {
+      // when both are present, for now,
+      // default to deno.json
+      (Some(deno), Some(_) | None) => {
+        Ok((ConfigFile::Deno(deno.clone()), factory))
+      }
+      (None, Some(package_json)) if options.enable_future_features() => {
+        Ok((ConfigFile::Npm(package_json.clone(), None), factory))
+      }
+      (None, Some(_) | None) => {
+        tokio::fs::write(options.initial_cwd().join("deno.json"), "{}\n")
+          .await
+          .context("Failed to create deno.json file")?;
+        log::info!("Created deno.json configuration file.");
+        let new_factory = CliFactory::from_flags(flags.clone())?;
+        let new_options = new_factory.cli_options().clone();
+        Ok((
+          ConfigFile::Deno(
+            new_options
+              .maybe_config_file()
+              .as_ref()
+              .ok_or_else(|| {
+                anyhow!("config not found, but it was just created")
+              })?
+              .clone(),
+          ),
+          new_factory,
+        ))
+      }
+    }
+  }
+}
+
+/// removes the `npm:` or `jsr:` prefix from a
+/// specifier (e.g. for inserting into
+/// `package.json`)
+fn strip_specifier_prefix(specifier: &str) -> &str {
+  let specifier = specifier.strip_prefix("npm:").unwrap_or(specifier);
+  specifier.strip_prefix("jsr:").unwrap_or(specifier)
+}
+
+fn package_json_dependency_entry(
+  selected: SelectedPackage,
+) -> (String, String) {
+  if selected.package_name.starts_with("npm:") {
```

Comment:
This could be `if Some(npm_package) = selected.package_name.strip_prefix("npm:")`, then the `strip_specifier_prefix` function could be removed, I think.